cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -52,7 +52,7 @@
     if (ARROW_PREDICT_FALSE(!_s.ok())) {                                            \
       std::stringstream ss;                                                         \
       ss << __FILE__ << ":" << __LINE__ << " code: " << #s << "\n" << _s.message(); \
-      ctx->SetStatus(Status(_s.code(), ss.str()));                                  \
+      ctx->SetStatus(Status(_s.code(), ss.str(), s.detail()));                      \
```

Comment:
Wow, I didn't know `cast.cc` redefined its own error-handling macros...

Reply:
yeah.  Somehow I missed the same issue in Status, which took a lot of time to find :/

Reply:
Perhaps open a JIRA to factor those out?

Reply:
https://issues.apache.org/jira/browse/ARROW-5833